python/ray/remote_function.py

```diff
@@ -202,29 +140,21 @@ def f():
         """
 
         func_cls = self
-        new_runtime_env = parse_runtime_env(runtime_env)
-
-        options = dict(
-            num_returns=num_returns,
-            num_cpus=num_cpus,
-            num_gpus=num_gpus,
-            memory=memory,
-            object_store_memory=object_store_memory,
-            accelerator_type=accelerator_type,
-            resources=resources,
-            max_retries=max_retries,
-            retry_exceptions=retry_exceptions,
-            placement_group=placement_group,
-            placement_group_bundle_index=placement_group_bundle_index,
-            placement_group_capture_child_tasks=(placement_group_capture_child_tasks),
-            runtime_env=new_runtime_env,
-            name=name,
-            scheduling_strategy=scheduling_strategy,
-        )
+
+        # override original options
+        options_from_remote_decorator = self._option_dict.copy()
+        # max_calls could not be used in ".options()", we should remove it before
+        # merging options from '@ray.remote'.
+        options_from_remote_decorator.pop("max_calls", None)
+        new_options = {**options_from_remote_decorator, **option_dict}
```

Comment:
Actually now I'm thinking of it, seems it does not exactly match the current merge logic?

For example
@ray.remote(placement_group=pg)
def func():
   pass

func.option(placement_group="default").remote()

The old behavior is that we will use `pg` as the placement group, but the new logic will just use "default"?

Reply:
I looked into the code and docs, "default" is not supposed to be used by users, it is just a placeholder representing "not specified in .options()". This is because `None` cannot be used to indicating the missing of `placement_group` field, it is a valid option.

So the logic is still consistent with the original code, just the original code process missing keywords in a weird way

Reply:
It's actually not just about "default" pg: the original code cannot differentiate between user explicitly setting an option to default (e.g. None, "default") and user doesn't set the option, so it will try to use the option set via remote() if the options() one has the default value. With the new code, IIUC, we can now differentiate these two cases by the existence of the option key in the dict and will use the options() one as long as it's set by the user explicitly regardless what the value is. Although normally, users don't set the option to default value explicitly but it's still possible that some users do `f.options(some_option=None).remote()` and expect to use the value of some_option set via remote().

Reply:
I disagree with this kind of usage. It is meaningless to set a value to `None/"default"` just for reusing the original value - instead you avoid setting the value in `.options()`.

So to me in the original code, this is just a way to detect missing of a field instead of "reusing the original option".

Reply:
according to the discussion in slack, we keep the design of this PR